### PR TITLE
Fix slow rate

### PR DESCRIPTION
**Pull request by edwardEvans094**

### CODE REVIEW

File: app/services/CurrenciesService.js [hunk 1]
```diff
@@ -18,27 +18,111 @@ module.exports = BaseService.extends({
...
+          FROM rate7d
+          where block_timestamp < ?
+      );
+    `;
+    const lastParams = [hour18Ago];
+
+
+    const pointSql = `
+      SELECT AVG(mid_expected) as rate7d, quote_symbol FROM rate 
+      WHERE mid_expected > 0 AND block_timestamp >= ? GROUP BY quote_symbol, h6_seq
```
Comment (cyberliem) [inline app/services/CurrenciesService.js]: similar to rate_24h, we should inject quote_symbol IN (list of symbol).

File: app/services/CurrenciesService.js [hunk 2]
```diff
@@ -18,27 +18,111 @@ module.exports = BaseService.extends({
...
+        }
+      })
+
+      const CACHE_KEY = CacheInfo.CurrenciesAllRates.key;
+      const CACHE_TTL = CacheInfo.CurrenciesAllRates.TTL;
+      const redisCacheService = this.getService('RedisCacheService');
+      redisCacheService.setCacheByKey(CACHE_KEY, pairs, CACHE_TTL)
+      const endTime = new Date().getTime()
+      console.log(`______ saved to redis in ${endTime - startTime} ms, cache ${CACHE_TTL.ttl} s`)
+      return callback(null, pairs)
```
Comment (cyberliem) [inline app/services/CurrenciesService.js]: So we're changing function signature . Please make sure that this function is not used anywhere else except for CurrencyController

File: config/cache/info.js [hunk 3]
```diff
@@ -1,10 +1,10 @@
 const Const = require('../../app/common/Const');
-
+const WEEK_IN_MILISECONDS = Const.MINUTE_IN_MILLISECONDS * 60 * 24 * 7
 module.exports = {
   "CurrenciesAllRates": {
     key: "all-rates",
-    TTL: {ttl: 63 * Const.MINUTE_IN_MILLISECONDS},
-    TTLTool: {ttl: 15 * Const.MINUTE_IN_MILLISECONDS}
+    TTL: {ttl: WEEK_IN_MILISECONDS},
```
Comment (cyberliem) [inline config/cache/info.js]: In theory a result whether from blockchain or from Database should only be avail for 1 hour. Now we're not concerning the DB query with a user request anymore (we return nil if cache doesn't hit), hence changing this number to 1 WEEK does not make much sense. I think 60 Minutes is ok.

File: app/services/CurrenciesService.js [hunk 4]
```diff
@@ -18,27 +18,111 @@ module.exports = BaseService.extends({
...
+    const startTime = new Date().getTime()
+
 
     const db = this._getDbConnection();
     //const tradeAdapter = db.model().getSlaveAdapter();
     const rateAdapter = db.model('Rate7dModel').getSlaveAdapter();
 
     let pairs = {}
 
-    Object.keys(tokens).forEach(token => {
```
[the remark this answers is not in the document]
Comment (edwardEvans094) [reply app/services/CurrenciesService.js]: ok let me add where quote_symbol IN to query

File: app/services/CurrenciesService.js [hunk 5]
```diff
@@ -18,27 +18,111 @@ module.exports = BaseService.extends({
...
+    const hour30Ago = nowInSeconds - 30 * 60 * 60;
+    const hour18Ago = nowInSeconds - 18 * 60 * 60;
+    const weekAgo = nowInSeconds - DAY_IN_SECONDS * 7;
+
+
+    const lastSql = `
+      SELECT mid_expected as 'rate_24h', quote_symbol
+      FROM rate7d
+      WHERE block_timestamp IN (
+          SELECT MAX(block_timestamp)
```
[the remark this answers is not in the document]
Comment (edwardEvans094) [reply app/services/CurrenciesService.js]: okey, let me fix

File: app/services/CurrenciesService.js (hunk 1 above)
Comment (edwardEvans094) [reply app/services/CurrenciesService.js]: okey

File: app/services/CurrenciesService.js (hunk 2 above)
Comment (edwardEvans094) [reply app/services/CurrenciesService.js]: just only use by currencyController

File: app/controllers/CurrenciesController.js [hunk 6]
```diff
@@ -112,48 +112,42 @@ module.exports = AppController.extends({
...
-          p: []
-        };
-        token.points.forEach((p) => {
-          item.p.push(p.rate7d);
-        });
+        if (ret) {
+          const pack = JSON.parse(ret)
+          return callback(null, pack)
+        }
+        return callback(null, null)
```
[the remark this answers is not in the document]
Comment (edwardEvans094) [reply app/controllers/CurrenciesController.js]: we have check if(!data) above, no need to return error here. the purpose of this func is get cached data, we need to now when error from redis or no data

File: config/cache/info.js (hunk 3 above)
Comment (edwardEvans094) [reply config/cache/info.js]: okey, will change to 1 hour

**Pull request merged**